### ISSUE TITLE
TestCQLService.StartupRequest is flaky

### ISSUE THREAD

**Issue by ttyusupov** [yugabyte/yugabyte-db#4751]
Fails ~1 of 100 times in ASAN build:
```
ybd --remote asan --cxx-test cqlserver_cqlserver-test --gtest_filter TestCQLService.StartupRequest -n 100 -- -p 1
```

```
*** Check failure stack trace: ***
F0611 16:50:29.830094 16870 thread_restrictions.cc:71] Check failed: LoadTLS()->io_allowed Function marked as IO-only was called from a thread that disallows IO!  If this thread really should be allowed to make IO calls, adjust the call to yb::ThreadRestrictions::SetIOAllowed() in this thread's startup. Thread 16870 (name: "TabletServer_reactor", category: "TabletServer_reactor")
    @     0x7f4a1d11a5e8  yb::(anonymous namespace)::DumpStackTraceAndExit() (src/yb/util/logging.cc:166)
    @     0x7f4a1ba0ad21
    @     0x7f4a1ba0704f
    @     0x7f4a1ba09979
    @     0x7f4a1ba1d213
    @     0x7f4a1d2f4657  yb::ThreadRestrictions::AssertIOAllowed() (src/yb/util/thread_restrictions.cc:71)
    @     0x7f4a1d0e21d9  yb::PosixRandomAccessFile::Read(unsigned long, unsigned long, yb::Slice*, unsigned char*) const (src/yb/util/file_system_posix.cc:170)
    @     0x7f4a1d0be5a9  yb::env_util::ReadFully(yb::RandomAccessFile*, unsigned long, unsigned long, yb::Slice*, unsigned char*) (src/yb/util/env_util.cc:151)
    @     0x7f4a30348b5a  yb::log::ReadableLogSegment::ReadEntryHeader(long*, yb::log::ReadableLogSegment::EntryHeader*) (src/yb/consensus/log_util.cc:756)
    @     0x7f4a30344ec5  yb::log::ReadableLogSegment::ReadEntryHeaderAndBatch(long*, yb::faststring*, yb::log::LogEntryBatchPB*) (src/yb/consensus/log_util.cc:747)
    @     0x7f4a303c0c92  yb::log::LogReader::ReadBatchUsingIndexEntry(yb::log::LogIndexEntry const&, yb::faststring*, yb::log::LogEntryBatchPB*) const (src/yb/consensus/log_reader.cc:401)
    @     0x7f4a303c1b88  yb::log::LogReader::ReadReplicatesInRange(long, long, long, std::vector<shared_ptr<yb::consensus::ReplicateMsg>, std::allocator<shared_ptr<yb::consensus::ReplicateMsg> > >*) const (src/yb/consensus/log_reader.cc:446)
    @     0x7f4a30a02dc4  yb::consensus::LogCache::ReadOps(long, long, int) (src/yb/consensus/log_cache.cc:318)
    @     0x7f4a309990ee  yb::consensus::PeerMessageQueue::ReadFromLogCache(long, long, int, string const&) (src/yb/consensus/consensus_queue.cc:629)
    @     0x7f4a309968e4  yb::consensus::PeerMessageQueue::RequestForPeer(string const&, yb::consensus::ConsensusRequestPB*, yb::consensus::ReplicateMsgsHolder*, bool*, yb::consensus::RaftPeerPB_MemberType*, bool*) (src/yb/consensus/consensus_queue.cc:544)
    @     0x7f4a30976b84  yb::consensus::Peer::SendNextRequest(yb::consensus::RequestTriggerMode) (src/yb/consensus/consensus_peers.cc:215)
    @     0x7f4a3097928a  yb::consensus::Peer::ProcessResponse() (src/yb/consensus/consensus_peers.cc:409)
    @     0x7f4a21da5149  yb::rpc::OutboundCall::InvokeCallbackSync() (src/yb/rpc/outbound_call.cc:384)
    @     0x7f4a21da56ee  yb::rpc::InvokeCallbackTask::Done(yb::Status const&) (src/yb/rpc/outbound_call.cc:174)
    @     0x7f4a21ef14a4  yb::rpc::ThreadPool::Impl::Enqueue(yb::rpc::ThreadPoolTask*) (src/yb/rpc/thread_pool.cc:181)
    @     0x7f4a21dab81a  yb::rpc::OutboundCall::InvokeCallback() (src/yb/rpc/outbound_call.cc:369)
    @     0x7f4a21dabe1e  yb::rpc::OutboundCall::SetResponse(yb::rpc::CallResponse&&) (src/yb/rpc/outbound_call.cc:427)
    @     0x7f4a21d3ecfd  yb::rpc::Connection::HandleCallResponse(yb::rpc::CallData*) (src/yb/rpc/connection.cc:325)
    @     0x7f4a21f00131  yb::rpc::YBOutboundConnectionContext::HandleCall(shared_ptr<yb::rpc::Connection> const&, yb::rpc::CallData*) (src/yb/rpc/yb_rpc.cc:540)
    @     0x7f4a21f00181  non-virtual thunk to yb::rpc::YBOutboundConnectionContext::HandleCall(shared_ptr<yb::rpc::Connection> const&, yb::rpc::CallData*) (src/yb/rpc/yb_rpc.cc:0)
    @     0x7f4a21d2cacb  yb::rpc::BinaryCallParser::Parse(shared_ptr<yb::rpc::Connection> const&, boost::container::small_vector<iovec, 4ul, boost::container::new_allocator<iovec> > const&, yb::StronglyTypedBool<yb::rpc::ReadBufferFull_Tag>, shared_ptr<yb::MemTracker> const*) (src/yb/rpc/binary_call_parser.cc:154)
    @     0x7f4a21f00c41  yb::rpc::YBOutboundConnectionContext::ProcessCalls(shared_ptr<yb::rpc::Connection> const&, boost::container::small_vector<iovec, 4ul, boost::container::new_allocator<iovec> > const&, yb::StronglyTypedBool<yb::rpc::ReadBufferFull_Tag>) (src/yb/rpc/yb_rpc.cc:561)
    @     0x7f4a21d3e256  yb::rpc::Connection::ProcessReceived(boost::container::small_vector<iovec, 4ul, boost::container::new_allocator<iovec> > const&, yb::StronglyTypedBool<yb::rpc::ReadBufferFull_Tag>) (src/yb/rpc/connection.cc:287)
    @     0x7f4a21edd626  yb::rpc::TcpStream::TryProcessReceived() (src/yb/rpc/tcp_stream.cc:389)
    @     0x7f4a21edbbf5  yb::rpc::TcpStream::ReadHandler() (src/yb/rpc/tcp_stream.cc:318)
    @     0x7f4a21edb2ab  yb::rpc::TcpStream::Handler(ev::io&, int) (src/yb/rpc/tcp_stream.cc:260)
    @     0x7f4a1b40cb6a
    @     0x7f4a1b410c79
    @     0x7f4a21dfef6f  yb::rpc::Reactor::RunThread() (src/yb/rpc/reactor.cc:481)
    @     0x7f4a1d2da9c0  yb::Thread::SuperviseThread(void*) (src/yb/util/thread.cc:746)
    @     0x7f4a15fb8693  start_thread (/tmp/glibc-20181130-26094-cs1x60/glibc-2.23/nptl/pthread_create.c:333)
    @     0x7f4a15ae341c  (unknown) (sysdeps/unix/sysv/linux/x86_64/clone.S:109)
    @ 0xffffffffffffffff
```

**Comment by kriteshranjan** [comment]
AI validation notes (`yb-issue-4751-analysis.md`):

# Analysis: yugabyte/yugabyte-db#4751

Issue: "TestCQLService.StartupRequest is flaky" (kind/failing-test, opened 2020-06-11, no Jira link, no comments).
Reported signature (ASAN build, ~1/100 runs of `ybd --remote asan --cxx-test cqlserver_cqlserver-test --gtest_filter TestCQLService.StartupRequest -n 100 -- -p 1`):

```
thread_restrictions.cc:71] Check failed: LoadTLS()->io_allowed Function marked as IO-only was called from a thread that disallows IO!  ... Thread 16870 (name: "TabletServer_reactor", ...)
```

Stack (innermost first): `ThreadRestrictions::AssertIOAllowed` <- `PosixRandomAccessFile::Read` <- `env_util::ReadFully` <- `log::ReadableLogSegment::ReadEntryHeader` <- `LogReader::ReadReplicatesInRange` <- `consensus::LogCache::ReadOps` (log_cache.cc:318) <- `PeerMessageQueue::RequestForPeer` (consensus_queue.cc:544) <- `Peer::SendNextRequest` (consensus_peers.cc:215) <- `Peer::ProcessResponse` (consensus_peers.cc:409) <- `rpc::OutboundCall::InvokeCallbackSync` (outbound_call.cc:384) <- `InvokeCallbackTask::Done` (outbound_call.cc:174) <- `rpc::ThreadPool::Impl::Enqueue` (thread_pool.cc:181) <- `OutboundCall::InvokeCallback` (outbound_call.cc:369) <- `OutboundCall::SetResponse` <- `Connection::HandleCallResponse` <- ... <- `Reactor::RunThread`.

Component: DocDB (src/yb/rpc + src/yb/consensus); the CQL test is only the vehicle (its `YBTableTestBase` fixture runs a multi-tserver mini-cluster and tears it down while Raft peers still exchange UpdateConsensus).

Validation date: 2026-09-06 (UTC).

## Verdict
NOT-REPRODUCED — the failing mechanism was fixed upstream by fa6766630d (#4984, D9606, 2020-10-12) and hardened by 37c4289540 (#8438, D11595, 2021-05-15); both fixes are present at every current support tag and at master HEAD 002a8169, and a deterministic runtime proxy that drives the identical shutdown-time code path (rejected UpdateConsensus callback executed synchronously on `TabletServer_reactor` threads) produced 0 check failures / 0 FATALs on all five current images. Quick verdict: Fixed (historical CI flake; issue can be closed).

## Root cause
Era source (commit 123f4d8182, 2020-06-10, the last master commit before the report; every frame's line number matches the issue's stack):

- `src/yb/rpc/thread_pool.cc:177-183` (era): `ThreadPool::Impl::Enqueue` returns `false` when `closing_` is set and first calls `task->Done(shutdown_status_)` (`:181`). The messenger shuts its callback thread pools down before its reactors (`Messenger::Shutdown`: `ShutdownThreadPools()` precedes `reactor->StartShutdown()`; today `src/yb/rpc/messenger.cc:199` vs `:231`), so during `MiniCluster` teardown a response that arrives on a reactor thread in that window is rejected by the pool.
- `src/yb/rpc/outbound_call.cc:168-175` (era): `InvokeCallbackTask::Done` reacted to the rejection by logging a WARNING and calling `call_->InvokeCallbackSync()` (`:174`) — i.e. it ran the user callback right there, on the reactor thread, without telling the callback anything went wrong and without lifting the reactor thread's IO restriction.
- `src/yb/consensus/consensus_peers.cc:327-410` (era): the callback is `Peer::ProcessResponse`. It read only `controller_.status()` (`:331`), which was OK because a real response had arrived (`Connection::HandleCallResponse` -> `SetResponse`), so it went on to `queue_->ResponseFromPeer` and, when the follower still had more ops to receive (`more_pending`), called `SendNextRequest(RequestTriggerMode::kAlwaysSend)` inline (`:409`). `SendNextRequest` -> `PeerMessageQueue::RequestForPeer` -> `ReadFromLogCache` -> `LogCache::ReadOps`; when the needed ops were not in the in-memory cache, `LogReader::ReadReplicatesInRange` read the WAL segment from disk.
- `src/yb/util/thread_restrictions.cc:70-71` (era; today `:74-75`): reactor threads run with IO disallowed, and `PosixRandomAccessFile::Read` asserts `AssertIOAllowed()`. The CHECK exists only when `ENABLE_THREAD_RESTRICTIONS` is defined, which `src/yb/util/thread_restrictions.h:37-38` ties to `#ifndef NDEBUG` — hence debug/ASAN-only, and flaky because it needs (a) a successful UpdateConsensus response landing in the thread-pool-closed-but-reactor-alive window, (b) a lagging follower (`more_pending`), and (c) a log-cache miss.

Fix 1 — fa6766630d22e9e9d7f0b83583548a0ba006366e ("[#4984] Various shutdown related fixes", 2020-10-12, D9606; its summary names exactly this: "During shutdown UpdateConsensus callback could be invoked in reactor thread so it will fail AssertIOAllowed check"): `InvokeCallbackTask::Done` now records the rejection on the call (`call_->SetThreadPoolFailure(status)`), exposed as `RpcController::thread_pool_failure()`, and `Peer::ProcessResponse` treats it as the call status (`if (status.ok()) status = controller_.thread_pool_failure();`), so the rejected callback takes `ProcessResponseError` and never reaches `SendNextRequest`/`LogCache::ReadOps`.

Fix 2 — 37c4289540d6a27bf56bf10b12375f2491d2e574 ("[#8438, #8439, #8450, #8395] Fix some TSAN tests", 2021-05-15, D11595): the same `Done` branch now calls `ThreadRestrictions::SetWaitAllowed(true)` and `ThreadRestrictions::SetIOAllowed(true)` before invoking the callback synchronously ("We are in the shutdown path, with the threadpool closing, so allow IO and wait"), so any consumer that still performs IO from a rejected callback cannot trip the CHECK.

Both fixes at the current targets (line numbers of `Peer::ProcessResponse` / `thread_pool_failure()` check / `InvokeCallbackTask::Done` / `SetIOAllowed(true)`):

| revision | consensus_peers.cc | outbound_call.cc | Enqueue rejection path |
|---|---|---|---|
| v2024.2.11.0 (04544fd9) | `:503` / `:508` | `:170` / `:179` | `src/yb/rpc/thread_pool.cc:56` `task->Done(status)` |
| v2025.1.4.0 (690ac414) | `:511` / `:516` | `:170` / `:179` | `src/yb/rpc/thread_pool.cc:65` |
| v2025.2.6.0 (ae22ef35) | `:509` / `:514` | `:196` / `:205` | `src/yb/util/thread_pool.cc:308-312` `task->Done(kShuttingDownStatus)` |
| v2026.1.1.1 (3b010e02) | `:509` / `:514` | `:199` / `:208` | `src/yb/util/thread_pool.cc:413-417` |
| v2.25.2.0 (03bbc9f3, preview) | `:504` / `:509` | `:166` / `:175` | `src/yb/rpc/thread_pool.cc:65` |
| master 002a8169830dad32a84a7f640aef18b40ec43d7f | `:509` / `:514` | `:199` / `:208` | `src/yb/util/thread_pool.cc:413-417` (+ `:437` worker-start failure, same `Done` handling) |

(master blobs fetched from GitHub raw at that SHA; the local clone is shallow for master.) `thread_restrictions.h:37-38` still gates the CHECK on `#ifndef NDEBUG` everywhere, so no shipped release binary can emit it. Still-affected targets: none. No master fix PR is warranted. Secondary observation (not part of this issue): the off-by-default multi-raft heartbeat batcher (`src/yb/consensus/multi_raft_batcher.cc:143-152`, `enable_multi_raft_heartbeat_batcher=false`) passes `controller.status()` to `Peer::ProcessHeartbeatResponse` without consulting `thread_pool_failure()`; it is protected only by Fix 2's `SetIOAllowed(true)`.

## Trigger matrix
Predictions were written before the runs. Release images compile the CHECK out, so the runtime cells assert (1) the binaries lack the CHECK message and contain the fixed `Done` branch, (2) the issue's code path — an UpdateConsensus response callback rejected by the closing pool and executed synchronously on a reactor thread — is actually exercised at graceful shutdown, and (3) nothing fails.

| condition | expected | observed |
|---|---|---|
| Era source 123f4d8182 (2020-06-10): `Peer::ProcessResponse` checks `thread_pool_failure()`? `Done` lifts IO restriction? | no / no (defect shape; positive control) | `consensus_peers.cc:331` reads only `controller_.status()`, `:409` inline `SendNextRequest`; `outbound_call.cc:174` bare `InvokeCallbackSync()`; `thread_pool.cc:181` `task->Done(shutdown_status_)` — all match the issue stack |
| Same two checks at v2024.2.11.0, v2025.1.4.0, v2025.2.6.0, v2026.1.1.1, v2.25.2.0, master 002a8169 | yes / yes (fixed) | yes / yes at every revision (table above) |
| Release binaries contain the CHECK text "Function marked as IO-only was called from a thread that" | 0 occurrences (compiled out under NDEBUG) | 0 in yb-tserver, yb-master and every `lib/yb/*.so` on all five images |
| Release binaries contain "Failed to schedule invoking callback on response for request" (fixed `Done` branch) | >= 1 | 3 per image (yb-tserver, yb-master, libyrpc.so) on all five images |
| Graceful stop of a leader tserver (t1, then t2) while UpdateConsensus calls to a SIGSTOPped follower are in flight | >= 1 rejected UpdateConsensus callback executed synchronously on a `TabletServer_reactor` thread; 0 `Check failed`; 0 `io_allowed`/`IO-only`; 0 FATAL; exit 0; "shut down complete" logged | t1/t2 rejected UpdateConsensus callbacks per image: 13/17 (2024.2), 14/16 (2025.1), 17/13 (2025.2), 16/14 (2026.1), 17/13 (2.25.2); every WARNING TID maps to a `TabletServer_re[actor]` thread (6 reactor threads each); 0/0/0 failures; exit 0; bye=1 — on all five images |
| Control: resume the paused follower (SIGCONT) and stop it gracefully | 0 rejected callbacks; clean exit | 0 rejected callbacks, 0 failures, exit 0, bye=1 on all five images |
| Consensus path positive evidence | UpdateConsensus handler counts grow across phases | ~470-500 per tserver after 300 rows; 1150-1900 on t1/t2 during the paused phase; 300 -> 600 rows read back on all five images |
| Graceful SIGTERM without the per-version switch (2024.2/2025.1/2025.2/2.25.2 first pass) | immediate exit 143, no `TabletServer::Shutdown` (`termination_monitor.cc:74-75` gating) | exit 143, bye=0, 0 rejected callbacks on those four; 2026.1 (default `graceful_shutdown=true`) already exit 0 with 12/18 rejected callbacks |

## Reproduction
Original repro needs an ASAN/debug build of the 2020-06 tree (`ybd --remote asan --cxx-test cqlserver_cqlserver-test --gtest_filter TestCQLService.StartupRequest -n 100 -- -p 1`); it cannot fire on release images because `AssertIOAllowed` is compiled out (`src/yb/util/thread_restrictions.h:37-38,127`). The executable proxy below drives the same shutdown-time path on the shipped binaries.

Scripts: `/tmp/claude/yb4751/yb4751-probe.sh.txt` (runs inside the image), driver `/tmp/claude/yb4751/run_matrix.sh`, combined log `/tmp/claude/yb4751/yb4751-matrix.log.txt`.

```sh
# host (Apple Silicon; podman machine mounts /private, so bind-mount via /private/tmp)
D=/tmp/claude/yb4751; TAG=2026.1.1.1-b2-aarch64      # one of the five tags below
podman run --timeout=900 --rm --name yb4751-$TAG -e EXTRA_TS_FLAGS="<see below>" \
  -v /private$D/yb4751-probe.sh.txt:/probe.sh:ro -v /private$D/out-$TAG:/out \
  docker.io/yugabytedb/yugabyte:$TAG bash /probe.sh
# EXTRA_TS_FLAGS: 2024.2.* / 2025.1.* / 2.25.* -> --TEST_running_test=true ; 2025.2.* -> --graceful_shutdown=true ; 2026.1.* -> (empty)
```

Probe steps (all loopback inside the container): `yb-tserver --version`; `grep -a -c` of the CHECK text and of the fixed-branch text over `/home/yugabyte/bin/yb-tserver`, `yb-master`, `/home/yugabyte/lib/yb/*.so`; 3 `yb-master` (127.0.0.1-3:7100, `--replication_factor=3 --enable_ysql=false`) + 3 `yb-tserver` (`--enable_ysql=false --yb_num_shards_per_tserver=2 --consensus_rpc_timeout_ms=20000` + `EXTRA_TS_FLAGS`); `ycqlsh`: `CREATE KEYSPACE k; CREATE TABLE k.t (id int PRIMARY KEY, v text) WITH transactions = {'enabled': true};` + 300 inserts; `kill -STOP` tserver 3; 300 more inserts (leaders now hold in-flight UpdateConsensus calls to t3 for up to 20 s); snapshot `/proc/<pid>/task/*/comm`; `kill -TERM` t1 and wait; same for t2; `kill -CONT` t3 then `kill -TERM` t3 (control); count in each tserver's `*log.INFO*`:

- failure signature: `Check failed`, `io_allowed`, `IO-only`, glog `F` lines, `*FATAL*` files (expect all 0);
- path-exercised signature: `Failed to schedule invoking callback on response for request yb.consensus.ConsensusService.UpdateConsensus to 127.0.0.3: Aborted (...): Service is shutting down` (expect >= 1 on t1 and t2, 0 on the control t3), and the emitting TIDs resolved to thread names (expect `TabletServer_re...`);
- `shut down complete` (expect 1) and exit status 0.

Observed on every image (details in the matrix log): `BINARY: check_msg_total=0 fix_msg_total=3`; t1/t2 `sched_fail_UpdateConsensus` 13-17 each from 6 `TabletServer_re` threads, `check_failed=0 io_allowed=0 fatal_lines=0 fatal_files=0 bye=1 exit=0`; control t3 `sched_fail_UpdateConsensus=0 ... exit=0`; master FATAL files 0. WARNING sites: `outbound_call.cc:173` (2024.2, 2025.1), `:199` (2025.2), `:202` (2026.1), `:169` (2.25.2) — each three lines below that tag's `InvokeCallbackTask::Done`, matching the source table.

## Versions tested
Resolved at: 2026-09-06T22:13Z (Docker Hub) / 2026-09-06T22:14Z (GitHub); runs 2026-09-06T22:29Z-22:34Z.
Resolution evidence: Docker Hub tags API (`ordering=last_updated`, `page_size=3`) — `2024.2*` count 87 newest `2024.2.11.0-b36(-aarch64)` pushed 2026-08-25T15:59Z; `2025.1*` count 36 newest `2025.1.4.0-b103` 2026-04-20T15:57Z; `2025.2*` count 39 newest `2025.2.6.0-b111` 2026-09-04T22:22Z; `2026.1*` count 12 newest `2026.1.1.1-b2` 2026-08-18T20:01Z (`latest` digest sha256:3926eedf...); `2.25.*` count 9 newest `2.25.2.0-b359` 2025-05-20T18:32Z; `2.27.`/`2.29.`/`2026.2` count 0. `github:list_commits` (`perPage=1`): v2024.2.11.0=04544fd94fb4e50d05994784f444b075a4a5dfdb, v2025.1.4.0=690ac4141e57d060fd834312f396643e8c41f260, v2025.2.6.0=ae22ef35ff65de63303467d95ac3e6d2382feba9, v2026.1.1.1=3b010e021f6027744bf9e016b17e108a2eb24e44, v2.25.2.0=03bbc9f31eae5c3fecabd8f0afa5e99211e6080d, master=002a8169830dad32a84a7f640aef18b40ec43d7f (committed 2026-09-06T19:35:46Z). Zero drift from the dispatcher's map. `yb-tserver --version` inside each image printed the same revision SHA and `build_type RELEASE`. Image digests are recorded in the matrix log.

| target | latest resolved version | source revision | observed repro/control | effective status |
|---|---|---|---|---|
| 2024.2 | 2024.2.11.0-b36 (image `2024.2.11.0-b36-aarch64`) | v2024.2.11.0 = 04544fd94fb4e50d05994784f444b075a4a5dfdb | CHECK text 0 / fix text 3; t1 13 + t2 17 rejected UpdateConsensus callbacks on reactor threads, 0 check failures, 0 FATAL, exit 0; control t3 0/clean (switch `--TEST_running_test=true`) | fixed |
| 2025.1 | 2025.1.4.0-b103 | v2025.1.4.0 = 690ac4141e57d060fd834312f396643e8c41f260 | 0 / 3; t1 14 + t2 16, 0 failures, exit 0; control clean (`--TEST_running_test=true`) | fixed |
| 2025.2 | 2025.2.6.0-b111 | v2025.2.6.0 = ae22ef35ff65de63303467d95ac3e6d2382feba9 | 0 / 3; t1 17 + t2 13, 0 failures, exit 0; control clean (`--graceful_shutdown=true`) | fixed |
| 2026.1 | 2026.1.1.1-b2 (= `latest`) | v2026.1.1.1 = 3b010e021f6027744bf9e016b17e108a2eb24e44 | 0 / 3; t1 16 + t2 14, 0 failures, exit 0; control clean (default `graceful_shutdown=true`); first pass 12/18 identical outcome | fixed |
| master | preview 2.25.2.0-b359 (runtime row) | image v2.25.2.0 = 03bbc9f31eae5c3fecabd8f0afa5e99211e6080d; source master HEAD 002a8169830dad32a84a7f640aef18b40ec43d7f (fix lines `consensus_peers.cc:514`, `outbound_call.cc:205-208`; HEAD postdates the preview by ~16 months, source-only) | 0 / 3; t1 17 + t2 13, 0 failures, exit 0; control clean (`--TEST_running_test=true`) | fixed |
| reported (2020-06 master, ASAN) | n/a (no release) | era commit 123f4d81829e98bd44e398658d795904ee1a15fe | defect shape confirmed in source (positive control); not runnable here | affected (historical) |

First released version containing the fix (tag ancestry in the local full-history clone; provenance only, not runtime proof): Fix 1 fa6766630d — v2.3.3.0 (2020-10-23, first tag) and stable v2.4.0 (2021-01-20); absent from every 2.2.x tag (last v2.2.7, 2021-02-02) and from v2.3.0.0. Fix 2 37c4289540 — v2.7.2 (2021-07-23) and stable v2.8.0 (2021-11-15); absent from 2.4.x and 2.6.x (v2.6.20.0 checked). Both are ancestors of v2024.2.11.0, v2025.1.4.0, v2026.1.1.1, v2.25.2.0 and v2025.2.0.0 (`git merge-base --is-ancestor`); v2025.2.6.0 is shallow locally, so its containment was confirmed by byte-presence of both fixes in its tree (table above). No backport is required on any in-support branch.

## Proposed fix direction
None required for the issue as reported: both fixes ship on every in-support branch and master. Recommended disposition: close #4751 as fixed by fa6766630d (#4984 / D9606) and 37c4289540 (#8438 / D11595). Optional hardening (separate issue, low value): make `MultiRaftHeartbeatBatcher`'s callback (`src/yb/consensus/multi_raft_batcher.cc:145`) fold `controller.thread_pool_failure()` into the status it hands to `Peer::ProcessHeartbeatResponse`, for symmetry with `Peer::ProcessResponse` (`consensus_peers.cc:512-515`); today that path is covered only by the reactor thread's `SetIOAllowed(true)` in `InvokeCallbackTask::Done`.

## Test strategy
Existing coverage: the original vehicle `TestCQLService.StartupRequest` (`src/yb/yql/cql/cqlserver/cqlserver-test.cc:196` at master HEAD) in a debug/ASAN build — `./yb_build.sh asan --cxx-test cqlserver_cqlserver-test --gtest_filter TestCQLService.StartupRequest -n 100` (CHECK only exists under `!NDEBUG`). Targeted unit coverage of the mechanism: an rpc test that shuts the messenger's callback thread pool down, completes an outbound call on the reactor thread and asserts (a) `RpcController::thread_pool_failure()` is non-OK in the callback and (b) `ThreadRestrictions::IsIOAllowed()` is true inside it (public `RpcController`/`ThreadRestrictions` APIs; `src/yb/rpc/rpc-test.cc` or `rpc_stub-test.cc` fixtures). Release-image regression check: the probe in this artifact (0 failure markers, >= 1 rejected UpdateConsensus callback on a reactor thread per graceful leader stop).

## Confidentiality check
Artifact, probe script and matrix log contain only public issue text, public commit SHAs/Phabricator IDs, public Docker Hub tags/digests, loopback addresses (127.0.0.1-3), synthetic keyspace/table names (`k.t`) and generic log lines. No customer data, PII, internal hostnames, credentials, or non-public information. No GitHub mutation was performed.

**Comment by kriteshranjan** [comment]
AI validation notes (`yb4751-probe.sh.txt`):

#!/bin/bash
# yb4751 probe (runs INSIDE a yugabytedb/yugabyte release image).
#
# Issue #4751: in a debug/ASAN test build, an UpdateConsensus response callback ran
# synchronously on a reactor thread because the RPC callback thread pool was already
# closing at shutdown; Peer::ProcessResponse then called SendNextRequest ->
# LogCache::ReadOps (WAL disk read) and tripped ThreadRestrictions::AssertIOAllowed.
#
# Release binaries compile AssertIOAllowed out, so this probe:
#   1. checks the shipped binaries: the CHECK message must be absent (compiled out) and
#      the fixed InvokeCallbackTask::Done branch message must be present;
#   2. runs a 3-master / 3-tserver RF3 cluster (YCQL only), writes rows so leaders exchange
#      UpdateConsensus with followers, pauses one follower (SIGSTOP) so UpdateConsensus
#      calls stay in flight, then gracefully stops the two live tservers so
#      Messenger::Shutdown rejects the in-flight callbacks (thread pools close before
#      reactors) -- the exact code path from the issue -- and counts the outcome.
#
# All addresses are loopback inside the container. Output goes to /out (bind mount).
set -u
B=/home/yugabyte/bin
OUT=/out
W=/tmp/yb4751
LOGS=$W/logs
MASTERS=127.0.0.1:7100,127.0.0.2:7100,127.0.0.3:7100
mkdir -p "$OUT" "$LOGS"
R="$OUT/result.txt"
: > "$R"
log() { echo "[$(date -u +%H:%M:%S)] $*" | tee -a "$R"; }

log "== version"
"$B/yb-tserver" --version 2>&1 | tee -a "$R"

log "== binary check (release build: CHECK message compiled out, fixed Done() branch present)"
CHECK_MSG='Function marked as IO-only was called from a thread that'
FIX_MSG='Failed to schedule invoking callback on response for request'
c_chk=0; c_fix=0
for f in "$B/yb-tserver" "$B/yb-master" /home/yugabyte/lib/yb/*.so; do
  [ -f "$f" ] || continue
  a=$(grep -a -c -F "$CHECK_MSG" "$f"); b=$(grep -a -c -F "$FIX_MSG" "$f")
  c_chk=$((c_chk + a)); c_fix=$((c_fix + b))
  if [ "$a" -gt 0 ] || [ "$b" -gt 0 ]; then log "   $f: check_msg=$a fix_msg=$b"; fi
done
log "BINARY: check_msg_total=$c_chk fix_msg_total=$c_fix"

log "== start 3 masters"
for i in 1 2 3; do
  mkdir -p "$W/m$i" "$LOGS/m$i"
  "$B/yb-master" --fs_data_dirs="$W/m$i" --master_addresses=$MASTERS \
    --rpc_bind_addresses=127.0.0.$i:7100 --webserver_interface=127.0.0.$i --webserver_port=7000 \
    --replication_factor=3 --enable_ysql=false --log_dir="$LOGS/m$i" \
    --default_memory_limit_to_ram_ratio=0.05 --yb_num_shards_per_tserver=2 \
    > "$LOGS/m$i.out" 2>&1 &
done
# Graceful SIGTERM handling (TerminationMonitor::InstallSigtermHandler) is gated per version:
#   2024.2 / 2025.1 / 2.25.x: only with --TEST_running_test=true (src/yb/common/termination_monitor.cc:74)
#   2025.2:                    --graceful_shutdown=true (default false; :75)
#   2026.1:                    --graceful_shutdown defaults to true (:77)
# The driver passes the matching switch in EXTRA_TS_FLAGS so TabletServer::Shutdown runs on every image.
EXTRA_TS_FLAGS="${EXTRA_TS_FLAGS:-}"
log "== start 3 tservers (consensus_rpc_timeout_ms=20000 keeps calls to a paused peer in flight; extra: '$EXTRA_TS_FLAGS')"
declare -A TPID
for i in 1 2 3; do
  mkdir -p "$W/t$i" "$LOGS/t$i"
  # shellcheck disable=SC2086
  "$B/yb-tserver" --fs_data_dirs="$W/t$i" --tserver_master_addrs=$MASTERS \
    --rpc_bind_addresses=127.0.0.$i:9100 --webserver_interface=127.0.0.$i --webserver_port=9000 \
    --cql_proxy_bind_address=127.0.0.$i:9042 --enable_ysql=false --start_redis_proxy=false \
    --log_dir="$LOGS/t$i" --default_memory_limit_to_ram_ratio=0.08 \
    --yb_num_shards_per_tserver=2 --consensus_rpc_timeout_ms=20000 $EXTRA_TS_FLAGS \
    > "$LOGS/t$i.out" 2>&1 &
  TPID[$i]=$!
done
log "tserver pids: ${TPID[1]} ${TPID[2]} ${TPID[3]}"

log "== wait for 3 live tservers"
for n in $(seq 1 90); do
  live=$(curl -s http://127.0.0.1:7000/api/v1/tablet-servers 2>/dev/null | grep -o '"status":"ALIVE"' | wc -l)
  [ "$live" -ge 3 ] && break
  sleep 2
done
log "live tservers: $live (after ~$((n*2))s)"

log "== create YCQL table (RF3, transactions enabled) and write rows"
"$B/ycqlsh" 127.0.0.1 9042 --request-timeout=60 -e "CREATE KEYSPACE IF NOT EXISTS k; CREATE TABLE IF NOT EXISTS k.t (id int PRIMARY KEY, v text) WITH transactions = {'enabled': true};" 2>&1 | tee -a "$R"
gen() { local from=$1 to=$2; for id in $(seq "$from" "$to"); do echo "INSERT INTO k.t (id, v) VALUES ($id, 'v$id');"; done; }
gen 1 300 > "$W/ins1.cql"
"$B/ycqlsh" 127.0.0.1 9042 --request-timeout=60 -f "$W/ins1.cql" 2>&1 | tail -3 | tee -a "$R"
cnt=$("$B/ycqlsh" 127.0.0.1 9042 --request-timeout=60 -e "SELECT COUNT(*) FROM k.t;" 2>&1 | grep -E '^\s*[0-9]+' | tr -d ' ')
log "rows after phase 1: $cnt"

# prometheus line format: name{labels} value timestamp_ms -> value is $(NF-1)
uc() { curl -s "http://127.0.0.$1:9000/prometheus-metrics" 2>/dev/null | awk '/^handler_latency_yb_consensus_ConsensusService_UpdateConsensus_count/ {s+=$(NF-1)} END {print s+0}'; }
log "UpdateConsensus handled (t1,t2,t3) before pause: $(uc 1) $(uc 2) $(uc 3)"

log "== pause t3 (SIGSTOP) so leaders keep UpdateConsensus calls to it in flight"
kill -STOP "${TPID[3]}"
sleep 8
gen 301 600 > "$W/ins2.cql"
"$B/ycqlsh" 127.0.0.1 9042 --request-timeout=60 -f "$W/ins2.cql" 2>&1 | tail -3 | tee -a "$R"
cnt=$("$B/ycqlsh" 127.0.0.1 9042 --request-timeout=60 -e "SELECT COUNT(*) FROM k.t;" 2>&1 | grep -E '^\s*[0-9]+' | tr -d ' ')
log "rows after phase 2 (t3 paused): $cnt"
log "UpdateConsensus handled (t1,t2) with t3 paused: $(uc 1) $(uc 2)"
sleep 2

stop_and_count() {
  local i=$1 pid=$2 t0 rc dur
  # Snapshot TID -> thread name before shutdown so glog thread ids in WARNING lines can be
  # attributed (the issue's stack ran on a "TabletServer_reactor" thread).
  for task in /proc/$pid/task/*; do echo "$(basename "$task") $(cat "$task/comm" 2>/dev/null)"; done > "$W/tids-t$i.txt"
  t0=$(date +%s)
  kill -TERM "$pid"
  for n in $(seq 1 120); do kill -0 "$pid" 2>/dev/null || break; sleep 1; done
  wait "$pid" 2>/dev/null; rc=$?
  dur=$(( $(date +%s) - t0 ))
  local inf="$LOGS/t$i"
  local sched sched_uc chk io fatal_lines fatal_files bye
  sched=$(cat "$inf"/*log.INFO* 2>/dev/null | grep -c -F "$FIX_MSG")
  sched_uc=$(cat "$inf"/*log.INFO* 2>/dev/null | grep -F "$FIX_MSG" | grep -c 'UpdateConsensus')
  chk=$(cat "$inf"/*log.INFO* "$LOGS/t$i.out" 2>/dev/null | grep -c 'Check failed')
  io=$(cat "$inf"/*log.INFO* "$LOGS/t$i.out" 2>/dev/null | grep -c -E 'io_allowed|IO-only')
  fatal_lines=$(cat "$inf"/*log.INFO* "$LOGS/t$i.out" 2>/dev/null | grep -c -E '^F[0-9]{4} ')
  fatal_files=$(ls "$inf"/*FATAL* 2>/dev/null | wc -l)
  bye=$(cat "$inf"/*log.INFO* 2>/dev/null | grep -c 'shut down complete')
  log "T$i graceful stop: exit=$rc after ${dur}s | sched_fail_total=$sched sched_fail_UpdateConsensus=$sched_uc | check_failed=$chk io_allowed=$io fatal_lines=$fatal_lines fatal_files=$fatal_files bye=$bye"
  cat "$inf"/*log.INFO* 2>/dev/null | grep -F "$FIX_MSG" | head -2 | sed 's/^/     /' | tee -a "$R"
  # Which threads emitted the rejected-callback WARNING (glog field 3 = Linux TID)?
  local names
  names=$(cat "$inf"/*log.INFO* 2>/dev/null | grep -F "$FIX_MSG" | awk '{print $3}' | sort -u | while read -r tid; do awk -v t="$tid" '$1==t {print $2}' "$W/tids-t$i.txt"; done | sort | uniq -c | awk '{printf "%s x%s; ", $2, $1}')
  log "T$i threads that ran the rejected callback synchronously: ${names:-none}"
}

log "== graceful stop of t1 (leader for some tablets) with t3 paused"
stop_and_count 1 "${TPID[1]}"
log "== graceful stop of t2 with t3 paused"
stop_and_count 2 "${TPID[2]}"

log "== resume t3 and stop it gracefully (control: no in-flight rejected callbacks expected to matter)"
kill -CONT "${TPID[3]}"
sleep 2
stop_and_count 3 "${TPID[3]}"

log "== stop masters"
pkill -TERM -f "$B/yb-master" ; sleep 3
m_fatal=$(ls "$LOGS"/m*/*FATAL* 2>/dev/null | wc -l)
log "master FATAL files: $m_fatal"
cp -r "$LOGS" "$OUT/logs" 2>/dev/null
log "== done"

**Comment by kriteshranjan** [comment]
AI validation notes (`yb4751-matrix.log.txt`):

# yugabyte/yugabyte-db#4751 runtime matrix (2026-09-06, podman arm64, docker.io/yugabytedb/yugabyte images)
# Probe: yb4751-probe.sh.txt (3 masters + 3 tservers, RF3 YCQL table, 600 rows, follower t3 paused with SIGSTOP,
#        graceful SIGTERM of t1 then t2 while UpdateConsensus calls to t3 are in flight; t3 resumed and stopped = control).
# Assertion per image: check_msg_total=0 (AssertIOAllowed CHECK compiled out of release binaries), fix_msg_total>=1
#   (InvokeCallbackTask::Done failure branch present), sched_fail_UpdateConsensus>=1 on t1/t2 (the issue's code path ran,
#   on TabletServer_reactor threads), check_failed=0 io_allowed=0 fatal_lines=0 fatal_files=0 bye=1 exit=0.
# Per-version switch installing the graceful SIGTERM handler (src/yb/common/termination_monitor.cc):
#   2024.2/2025.1/2.25.x --TEST_running_test=true (:74) | 2025.2 --graceful_shutdown=true (:75) | 2026.1 default true (:77)
#
# Resolution evidence (Docker Hub tags API, ordering=last_updated, queried 2026-09-06T22:1xZ):
#   2024.2   count=87   newest: 2024.2.11.0-b36@2026-08-25T16:02:12Z | 2024.2.11.0-b36-x86_64@2026-08-25T16:02:08Z | 2024.2.11.0-b36-aarch64@2026-08-25T15:59:27Z
#   2025.1   count=36   newest: 2025.1.4.0-b103@2026-04-20T15:59:42Z | 2025.1.4.0-b103-x86_64@2026-04-20T15:59:39Z | 2025.1.4.0-b103-aarch64@2026-04-20T15:57:55Z
#   2025.2   count=39   newest: 2025.2.6.0-b111@2026-09-04T22:23:02Z | 2025.2.6.0-b111-x86_64@2026-09-04T22:22:58Z | 2025.2.6.0-b111-aarch64@2026-09-04T22:22:00Z
#   2026.1   count=12   newest: 2026.1.1.1-b2@2026-08-18T20:03:16Z | 2026.1.1.1-b2-x86_64@2026-08-18T20:03:12Z | 2026.1.1.1-b2-aarch64@2026-08-18T20:01:48Z
#   2.25.    count=9    newest: 2.25.2.0-b359@2025-05-20T18:34:42Z | 2.25.2.0-b359-x86_64@2025-05-20T18:34:39Z | 2.25.2.0-b359-aarch64@2025-05-20T18:32:55Z
#   2.27.    count=0    newest: 
#   2.29.    count=0    newest: 
#   2026.2   count=0    newest: 
#   tag SHAs via github:list_commits: v2024.2.11.0=04544fd94fb4e50d05994784f444b075a4a5dfdb v2025.1.4.0=690ac4141e57d060fd834312f396643e8c41f260
#   v2025.2.6.0=ae22ef35ff65de63303467d95ac3e6d2382feba9 v2026.1.1.1=3b010e021f6027744bf9e016b17e108a2eb24e44 v2.25.2.0=03bbc9f31eae5c3fecabd8f0afa5e99211e6080d
#   master HEAD=002a8169830dad32a84a7f640aef18b40ec43d7f (committed 2026-09-06T19:35:46Z)
# Image digests (podman image inspect):
#   2024.2.11.0-b36-aarch64    sha256:53c01c56315e245cc901cbc46e93f30fedb1f87833c7b1aa09d20fbeda27a288
#   2025.1.4.0-b103-aarch64    sha256:08f1be0d6e074e0286e70067c12933a8cf536876d50fc43c52b1127d447ecdc0
#   2025.2.6.0-b111-aarch64    sha256:4a1f7a94d3b70c61ecaf5e049521d691b16c9bf204ed01bdb6ab75d5beeb623e
#   2026.1.1.1-b2-aarch64      sha256:d27eede72bbc669e348e9920318589b4bb68e98204a241ec6bfaba04bf4f6699
#   2.25.2.0-b359-aarch64      sha256:3f9cd1b8f520cc78e82ace51174be73343a551148bc6c9a9f821a22f1d2858c5
# Fix provenance (tag ancestry, local clone): fa6766630d22e9e9d7f0b83583548a0ba006366e (#4984, D9606, 2020-10-12) first in v2.3.3.0 / v2.4.0;
#   37c4289540d6a27bf56bf10b12375f2491d2e574 (#8438 et al., D11595, 2021-05-15) first in v2.7.2 / v2.8.0. Never in 2.2.x (v2.2.7 last).

################ image docker.io/yugabytedb/yugabyte:2024.2.11.0-b36-aarch64  (EXTRA_TS_FLAGS='--TEST_running_test=true')
[22:29:47] == version
version 2024.2.11.0 build 36 revision 04544fd94fb4e50d05994784f444b075a4a5dfdb build_type RELEASE built at 18 Aug 2026 00:58:07 UTC
[22:29:47] == binary check (release build: CHECK message compiled out, fixed Done() branch present)
[22:29:48]    /home/yugabyte/bin/yb-tserver: check_msg=0 fix_msg=1
[22:29:48]    /home/yugabyte/bin/yb-master: check_msg=0 fix_msg=1
[22:29:49]    /home/yugabyte/lib/yb/libyrpc.so: check_msg=0 fix_msg=1
[22:29:49] BINARY: check_msg_total=0 fix_msg_total=3
[22:29:49] == start 3 masters
[22:29:49] == start 3 tservers (consensus_rpc_timeout_ms=20000 keeps calls to a paused peer in flight; extra: '--TEST_running_test=true')
[22:29:49] tserver pids: 207 209 218
[22:29:49] == wait for 3 live tservers
[22:29:53] live tservers: 3 (after ~6s)
[22:29:53] == create YCQL table (RF3, transactions enabled) and write rows
[22:29:55] rows after phase 1: 300
[22:29:55] UpdateConsensus handled (t1,t2,t3) before pause: 487 501 473
[22:29:55] == pause t3 (SIGSTOP) so leaders keep UpdateConsensus calls to it in flight
[22:30:25] rows after phase 2 (t3 paused): 600
[22:30:25] UpdateConsensus handled (t1,t2) with t3 paused: 1909 1450
[22:30:27] == graceful stop of t1 (leader for some tablets) with t3 paused
[22:30:29] T1 graceful stop: exit=0 after 1s | sched_fail_total=13 sched_fail_UpdateConsensus=13 | check_failed=0 io_allowed=0 fatal_lines=0 fatal_files=0 bye=1
     W0906 22:30:28.205925   307 outbound_call.cc:173] Failed to schedule invoking callback on response for request yb.consensus.ConsensusService.UpdateConsensus to 127.0.0.3: Aborted (yb/rpc/thread_pool.cc:50): Service is shutting down
     W0906 22:30:28.205947   312 outbound_call.cc:173] Failed to schedule invoking callback on response for request yb.consensus.ConsensusService.UpdateConsensus to 127.0.0.3: Aborted (yb/rpc/thread_pool.cc:50): Service is shutting down
[22:30:29] T1 threads that ran the rejected callback synchronously: TabletServer_re x6; 
[22:30:29] == graceful stop of t2 with t3 paused
[22:30:30] T2 graceful stop: exit=0 after 1s | sched_fail_total=17 sched_fail_UpdateConsensus=17 | check_failed=0 io_allowed=0 fatal_lines=0 fatal_files=0 bye=1
     W0906 22:30:29.814864   324 outbound_call.cc:173] Failed to schedule invoking callback on response for request yb.consensus.ConsensusService.UpdateConsensus to 127.0.0.3: Aborted (yb/rpc/thread_pool.cc:50): Service is shutting down
     W0906 22:30:29.815056   324 outbound_call.cc:173] Failed to schedule invoking callback on response for request yb.consensus.ConsensusService.UpdateConsensus to 127.0.0.3: Aborted (yb/rpc/thread_pool.cc:50): Service is shutting down
[22:30:30] T2 threads that ran the rejected callback synchronously: TabletServer_re x6; 
[22:30:30] == resume t3 and stop it gracefully (control: no in-flight rejected callbacks expected to matter)
[22:30:34] T3 graceful stop: exit=0 after 1s | sched_fail_total=0 sched_fail_UpdateConsensus=0 | check_failed=0 io_allowed=0 fatal_lines=0 fatal_files=0 bye=1
[22:30:34] T3 threads that ran the rejected callback synchronously: none
[22:30:34] == stop masters
[22:30:37] master FATAL files: 0
[22:30:37] == done

################ image docker.io/yugabytedb/yugabyte:2025.1.4.0-b103-aarch64  (EXTRA_TS_FLAGS='--TEST_running_test=true')
[22:30:45] == version
version 2025.1.4.0 build 103 revision 690ac4141e57d060fd834312f396643e8c41f260 build_type RELEASE built at 02 Apr 2026 20:43:14 UTC
[22:30:45] == binary check (release build: CHECK message compiled out, fixed Done() branch present)
[22:30:46]    /home/yugabyte/bin/yb-tserver: check_msg=0 fix_msg=1
[22:30:46]    /home/yugabyte/bin/yb-master: check_msg=0 fix_msg=1
[22:30:47]    /home/yugabyte/lib/yb/libyrpc.so: check_msg=0 fix_msg=1
[22:30:47] BINARY: check_msg_total=0 fix_msg_total=3
[22:30:47] == start 3 masters
[22:30:47] == start 3 tservers (consensus_rpc_timeout_ms=20000 keeps calls to a paused peer in flight; extra: '--TEST_running_test=true')
[22:30:47] tserver pids: 189 194 203
[22:30:47] == wait for 3 live tservers
[22:30:51] live tservers: 3 (after ~6s)
[22:30:51] == create YCQL table (RF3, transactions enabled) and write rows
[22:30:52] rows after phase 1: 300
[22:30:53] UpdateConsensus handled (t1,t2,t3) before pause: 469 447 396
[22:30:53] == pause t3 (SIGSTOP) so leaders keep UpdateConsensus calls to it in flight
[22:31:08] rows after phase 2 (t3 paused): 600
[22:31:08] UpdateConsensus handled (t1,t2) with t3 paused: 1238 1154
[22:31:10] == graceful stop of t1 (leader for some tablets) with t3 paused
[22:31:11] T1 graceful stop: exit=0 after 1s | sched_fail_total=14 sched_fail_UpdateConsensus=14 | check_failed=0 io_allowed=0 fatal_lines=0 fatal_files=0 bye=1
     W0906 22:31:10.954246   290 outbound_call.cc:173] Failed to schedule invoking callback on response for request yb.consensus.ConsensusService.UpdateConsensus to 127.0.0.3: Aborted (yb/rpc/thread_pool.cc:59): Service is shutting down
     W0906 22:31:10.954437   290 outbound_call.cc:173] Failed to schedule invoking callback on response for request yb.consensus.ConsensusService.UpdateConsensus to 127.0.0.3: Aborted (yb/rpc/thread_pool.cc:59): Service is shutting down
[22:31:11] T1 threads that ran the rejected callback synchronously: TabletServer_re x6; 
[22:31:11] == graceful stop of t2 with t3 paused
[22:31:13] T2 graceful stop: exit=0 after 1s | sched_fail_total=16 sched_fail_UpdateConsensus=16 | check_failed=0 io_allowed=0 fatal_lines=0 fatal_files=0 bye=1
     W0906 22:31:12.362558   284 outbound_call.cc:173] Failed to schedule invoking callback on response for request yb.consensus.ConsensusService.UpdateConsensus to 127.0.0.3: Aborted (yb/rpc/thread_pool.cc:59): Service is shutting down
     W0906 22:31:12.362553   277 outbound_call.cc:173] Failed to schedule invoking callback on response for request yb.consensus.ConsensusService.UpdateConsensus to 127.0.0.3: Aborted (yb/rpc/thread_pool.cc:59): Service is shutting down
[22:31:13] T2 threads that ran the rejected callback synchronously: TabletServer_re x6; 
[22:31:13] == resume t3 and stop it gracefully (control: no in-flight rejected callbacks expected to matter)
[22:31:17] T3 graceful stop: exit=0 after 1s | sched_fail_total=0 sched_fail_UpdateConsensus=0 | check_failed=0 io_allowed=0 fatal_lines=0 fatal_files=0 bye=1
[22:31:17] T3 threads that ran the rejected callback synchronously: none
[22:31:17] == stop masters
[22:31:20] master FATAL files: 0
[22:31:20] == done

################ image docker.io/yugabytedb/yugabyte:2025.2.6.0-b111-aarch64  (EXTRA_TS_FLAGS='--graceful_shutdown=true')
[22:31:27] == version
version 2025.2.6.0 build 111 revision ae22ef35ff65de63303467d95ac3e6d2382feba9 build_type RELEASE built at 02 Sep 2026 16:42:37 UTC
[22:31:27] == binary check (release build: CHECK message compiled out, fixed Done() branch present)
[22:31:27]    /home/yugabyte/bin/yb-tserver: check_msg=0 fix_msg=1
[22:31:27]    /home/yugabyte/bin/yb-master: check_msg=0 fix_msg=1
[22:31:28]    /home/yugabyte/lib/yb/libyrpc.so: check_msg=0 fix_msg=1
[22:31:28] BINARY: check_msg_total=0 fix_msg_total=3
[22:31:28] == start 3 masters
[22:31:28] == start 3 tservers (consensus_rpc_timeout_ms=20000 keeps calls to a paused peer in flight; extra: '--graceful_shutdown=true')
[22:31:28] tserver pids: 193 195 207
[22:31:28] == wait for 3 live tservers
[22:31:32] live tservers: 3 (after ~6s)
[22:31:32] == create YCQL table (RF3, transactions enabled) and write rows
[22:31:33] rows after phase 1: 300
[22:31:33] UpdateConsensus handled (t1,t2,t3) before pause: 492 460 350
[22:31:33] == pause t3 (SIGSTOP) so leaders keep UpdateConsensus calls to it in flight
[22:32:03] rows after phase 2 (t3 paused): 600
[22:32:03] UpdateConsensus handled (t1,t2) with t3 paused: 1448 1861
[22:32:05] == graceful stop of t1 (leader for some tablets) with t3 paused
[22:32:07] T1 graceful stop: exit=0 after 1s | sched_fail_total=17 sched_fail_UpdateConsensus=17 | check_failed=0 io_allowed=0 fatal_lines=0 fatal_files=0 bye=1
     W0906 22:32:06.509549   279 outbound_call.cc:199] Failed to schedule invoking callback on response for request yb.consensus.ConsensusService.UpdateConsensus to 127.0.0.3: Aborted (yb/util/thread_pool.cc:85): Service is shutting down
     W0906 22:32:06.509552   273 outbound_call.cc:199] Failed to schedule invoking callback on response for request yb.consensus.ConsensusService.UpdateConsensus to 127.0.0.3: Aborted (yb/util/thread_pool.cc:85): Service is shutting down
[22:32:07] T1 threads that ran the rejected callback synchronously: TabletServer_re x6; 
[22:32:07] == graceful stop of t2 with t3 paused
[22:32:08] T2 graceful stop: exit=0 after 1s | sched_fail_total=13 sched_fail_UpdateConsensus=13 | check_failed=0 io_allowed=0 fatal_lines=0 fatal_files=0 bye=1
     W0906 22:32:07.836493   299 outbound_call.cc:199] Failed to schedule invoking callback on response for request yb.consensus.ConsensusService.UpdateConsensus to 127.0.0.3: Aborted (yb/util/thread_pool.cc:85): Service is shutting down
     W0906 22:32:07.836509   305 outbound_call.cc:199] Failed to schedule invoking callback on response for request yb.consensus.ConsensusService.UpdateConsensus to 127.0.0.3: Aborted (yb/util/thread_pool.cc:85): Service is shutting down
[22:32:08] T2 threads that ran the rejected callback synchronously: TabletServer_re x6; 
[22:32:08] == resume t3 and stop it gracefully (control: no in-flight rejected callbacks expected to matter)
[22:32:12] T3 graceful stop: exit=0 after 1s | sched_fail_total=0 sched_fail_UpdateConsensus=0 | check_failed=0 io_allowed=0 fatal_lines=0 fatal_files=0 bye=1
[22:32:12] T3 threads that ran the rejected callback synchronously: none
[22:32:12] == stop masters
[22:32:15] master FATAL files: 0
[22:32:15] == done

################ image docker.io/yugabytedb/yugabyte:2026.1.1.1-b2-aarch64  (EXTRA_TS_FLAGS='none')
[22:32:23] == version
version 2026.1.1.1 build 2 revision 3b010e021f6027744bf9e016b17e108a2eb24e44 build_type RELEASE built at 14 Aug 2026 04:03:14 UTC
[22:32:23] == binary check (release build: CHECK message compiled out, fixed Done() branch present)
[22:32:23]    /home/yugabyte/bin/yb-tserver: check_msg=0 fix_msg=1
[22:32:23]    /home/yugabyte/bin/yb-master: check_msg=0 fix_msg=1
[22:32:24]    /home/yugabyte/lib/yb/libyrpc.so: check_msg=0 fix_msg=1
[22:32:24] BINARY: check_msg_total=0 fix_msg_total=3
[22:32:24] == start 3 masters
[22:32:24] == start 3 tservers (consensus_rpc_timeout_ms=20000 keeps calls to a paused peer in flight; extra: '')
[22:32:24] tserver pids: 191 193 206
[22:32:24] == wait for 3 live tservers
[22:32:28] live tservers: 3 (after ~6s)
[22:32:28] == create YCQL table (RF3, transactions enabled) and write rows
[22:32:29] rows after phase 1: 300
[22:32:29] UpdateConsensus handled (t1,t2,t3) before pause: 466 483 495
[22:32:29] == pause t3 (SIGSTOP) so leaders keep UpdateConsensus calls to it in flight
[22:33:00] rows after phase 2 (t3 paused): 600
[22:33:00] UpdateConsensus handled (t1,t2) with t3 paused: 1621 1690
[22:33:02] == graceful stop of t1 (leader for some tablets) with t3 paused
[22:33:03] T1 graceful stop: exit=0 after 1s | sched_fail_total=16 sched_fail_UpdateConsensus=16 | check_failed=0 io_allowed=0 fatal_lines=0 fatal_files=0 bye=1
     W0906 22:33:02.842427   278 outbound_call.cc:202] Failed to schedule invoking callback on response for request yb.consensus.ConsensusService.UpdateConsensus to 127.0.0.3: Aborted (yb/util/thread_pool.cc:103): Service is shutting down
     W0906 22:33:02.842530   271 outbound_call.cc:202] Failed to schedule invoking callback on response for request yb.consensus.ConsensusService.UpdateConsensus to 127.0.0.3: Aborted (yb/util/thread_pool.cc:103): Service is shutting down
[22:33:03] T1 threads that ran the rejected callback synchronously: TabletServer_re x6; 
[22:33:03] == graceful stop of t2 with t3 paused
[22:33:05] T2 graceful stop: exit=0 after 1s | sched_fail_total=14 sched_fail_UpdateConsensus=14 | check_failed=0 io_allowed=0 fatal_lines=0 fatal_files=0 bye=1
     W0906 22:33:04.195315   298 outbound_call.cc:202] Failed to schedule invoking callback on response for request yb.consensus.ConsensusService.UpdateConsensus to 127.0.0.3: Aborted (yb/util/thread_pool.cc:103): Service is shutting down
     W0906 22:33:04.195456   298 outbound_call.cc:202] Failed to schedule invoking callback on response for request yb.consensus.ConsensusService.UpdateConsensus to 127.0.0.3: Aborted (yb/util/thread_pool.cc:103): Service is shutting down
[22:33:05] T2 threads that ran the rejected callback synchronously: TabletServer_re x6; 
[22:33:05] == resume t3 and stop it gracefully (control: no in-flight rejected callbacks expected to matter)
[22:33:08] T3 graceful stop: exit=0 after 1s | sched_fail_total=0 sched_fail_UpdateConsensus=0 | check_failed=0 io_allowed=0 fatal_lines=0 fatal_files=0 bye=1
[22:33:08] T3 threads that ran the rejected callback synchronously: none
[22:33:08] == stop masters
[22:33:11] master FATAL files: 0
[22:33:11] == done

################ image docker.io/yugabytedb/yugabyte:2.25.2.0-b359-aarch64  (EXTRA_TS_FLAGS='--TEST_running_test=true')
[22:33:19] == version
version 2.25.2.0 build 359 revision 03bbc9f31eae5c3fecabd8f0afa5e99211e6080d build_type RELEASE built at 08 May 2025 19:05:35 UTC
[22:33:19] == binary check (release build: CHECK message compiled out, fixed Done() branch present)
[22:33:20]    /home/yugabyte/bin/yb-tserver: check_msg=0 fix_msg=1
[22:33:20]    /home/yugabyte/bin/yb-master: check_msg=0 fix_msg=1
[22:33:21]    /home/yugabyte/lib/yb/libyrpc.so: check_msg=0 fix_msg=1
[22:33:21] BINARY: check_msg_total=0 fix_msg_total=3
[22:33:21] == start 3 masters
[22:33:21] == start 3 tservers (consensus_rpc_timeout_ms=20000 keeps calls to a paused peer in flight; extra: '--TEST_running_test=true')
[22:33:21] tserver pids: 185 187 200
[22:33:21] == wait for 3 live tservers
[22:33:25] live tservers: 3 (after ~6s)
[22:33:25] == create YCQL table (RF3, transactions enabled) and write rows
[22:33:26] rows after phase 1: 300
[22:33:27] UpdateConsensus handled (t1,t2,t3) before pause: 496 482 495
[22:33:27] == pause t3 (SIGSTOP) so leaders keep UpdateConsensus calls to it in flight
[22:33:57] rows after phase 2 (t3 paused): 600
[22:33:57] UpdateConsensus handled (t1,t2) with t3 paused: 1540 1822
[22:33:59] == graceful stop of t1 (leader for some tablets) with t3 paused
[22:34:01] T1 graceful stop: exit=0 after 2s | sched_fail_total=17 sched_fail_UpdateConsensus=17 | check_failed=0 io_allowed=0 fatal_lines=0 fatal_files=0 bye=1
     W0906 22:34:00.025239   283 outbound_call.cc:169] Failed to schedule invoking callback on response for request yb.consensus.ConsensusService.UpdateConsensus to 127.0.0.3: Aborted (yb/rpc/thread_pool.cc:59): Service is shutting down
     W0906 22:34:00.025365   283 outbound_call.cc:169] Failed to schedule invoking callback on response for request yb.consensus.ConsensusService.UpdateConsensus to 127.0.0.3: Aborted (yb/rpc/thread_pool.cc:59): Service is shutting down
[22:34:01] T1 threads that ran the rejected callback synchronously: TabletServer_re x6; 
[22:34:01] == graceful stop of t2 with t3 paused
[22:34:02] T2 graceful stop: exit=0 after 1s | sched_fail_total=13 sched_fail_UpdateConsensus=13 | check_failed=0 io_allowed=0 fatal_lines=0 fatal_files=0 bye=1
     W0906 22:34:01.464046   287 outbound_call.cc:169] Failed to schedule invoking callback on response for request yb.consensus.ConsensusService.UpdateConsensus to 127.0.0.3: Aborted (yb/rpc/thread_pool.cc:59): Service is shutting down
     W0906 22:34:01.464051   288 outbound_call.cc:169] Failed to schedule invoking callback on response for request yb.consensus.ConsensusService.UpdateConsensus to 127.0.0.3: Aborted (yb/rpc/thread_pool.cc:59): Service is shutting down
[22:34:02] T2 threads that ran the rejected callback synchronously: TabletServer_re x6; 
[22:34:02] == resume t3 and stop it gracefully (control: no in-flight rejected callbacks expected to matter)
[22:34:06] T3 graceful stop: exit=0 after 1s | sched_fail_total=0 sched_fail_UpdateConsensus=0 | check_failed=0 io_allowed=0 fatal_lines=0 fatal_files=0 bye=1
[22:34:06] T3 threads that ran the rejected callback synchronously: none
[22:34:06] == stop masters
[22:34:09] master FATAL files: 0
[22:34:09] == done

**Comment by kriteshranjan** [comment]
Closing this issue based on AI validation. cc: @m-iancu , @VamsidharGunturu
